packages/server-wallet/src/wallet/wallet.ts

```diff
@@ -136,112 +251,138 @@ export class Wallet {
-
-        await this._messageService.send(messagesForObjective);
       }
-      if (isComplete) return {channelId: objective.data.targetChannelId, type: 'Success'};
-      return {numberOfAttempts: this._retryOptions.numberOfAttempts, type: 'EnsureObjectiveFailed'};
-    } catch (error) {
-      this._engine.logger.error({err: error}, 'Uncaught error in EnsureObjective');
-      return {
-        type: 'InternalError' as const,
-        error,
-      };
     }
   }
 
   /**
-   * Gets a message for a specific objective from a SyncObjectiveResult
-   * Throws if there are additional messages or the message is missing
+   * Emits events, sends messages and requests transactions based on the output of the engine.
+   * @param output
    */
-  private getMessagesForObjective(objectiveId: string, result: SyncObjectiveResult) {
-    const objectiveIds = Object.keys(result.messagesByObjective);
-    if (!objectiveIds.includes(objectiveId)) {
-      throw new Error(`No messages for objective ${objectiveId}`);
-
-      // This is a sanity check to prevent us from losing messages
-    } else if (objectiveIds.length !== 1) {
-      throw new Error(`There are messages for multiple objectives ${objectiveIds}`);
-    }
-
-    return result.messagesByObjective[objectiveId];
+  private async handleEngineOutput(
+    output: MultipleChannelOutput | SingleChannelOutput
+  ): Promise<void> {
+    this.emitObjectiveEvents(output);
+    await this._messageService.send(getMessages(output));
+    await this._chainService.handleChainRequests(output.chainRequests);
   }
 
   public get messageService(): MessageServiceInterface {
     return this._messageService;
   }
 
+  private async createObjectiveDoneResult(
+    objective: WalletObjective
+  ): Promise<ObjectiveDoneResult> {
+    return new Promise<ObjectiveDoneResult>(resolve => {
+      this.on('ObjectiveTimedOut', o => {
+        if (o.objectiveId === objective.objectiveId) {
+          this._engine.logger.trace({objective: o}, 'Objective Timed out');
+
+          resolve({
+            type: 'ObjectiveTimedOutError',
+            objectiveId: o.objectiveId,
+            lastProgressMadeAt: o.progressLastMadeAt,
+          });
+        }
+      });
+      this.on('ObjectiveCompleted', o => {
+        if (o.objectiveId === objective.objectiveId) {
+          this._engine.logger.trace({objective: o}, 'Objective Suceeded');
+          resolve({type: 'Success', channelId: o.data.targetChannelId});
+        }
+      });
+    });
+  }
+
+  /**
+   * Registers any channels existing in the database with the chain service.
+   *
+   * @remarks
+   * Enables the chain service to alert the engine of of any blockchain events for existing channels.
+   *
+   * @returns A promise that resolves when the channels have been successfully registered.
+   */
+  private async registerExistingChannelsWithChainService(): Promise<void> {
+    const channelsToRegister = (await this._engine.store.getNonFinalizedChannels()).map(
```

Comment:
unrelated to this PR, but `getNonFinalizedChannels` seems to only return `'Open'` channels. But:
- A channel in the `Challenge` mode is not finalized
- Should we also register channels that are _finalized_ but still have holdings on chain?